test/e2e/specs/editor/various/undo.spec.js

```diff
@@ -0,0 +1,516 @@
+/**
+ * WordPress dependencies
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.use( {
+	undoUtils: async ( { page }, use ) => {
+		await use( new UndoUtils( { page } ) );
+	},
+} );
+
+class UndoUtils {
+	constructor( { page } ) {
+		this.page = page;
+
+		this.getSelection = this.getSelection.bind( this );
+	}
+
+	async getSelection() {
+		return await this.page.evaluate( () => {
+			const selectedBlock = document.activeElement.closest( '.wp-block' );
+			const blocks = Array.from(
+				document.querySelectorAll( '.wp-block' )
+			);
+			const blockIndex = blocks.indexOf( selectedBlock );
+
+			if ( blockIndex === -1 ) {
+				return {};
+			}
+
+			let editables;
+
+			if ( selectedBlock.getAttribute( 'contenteditable' ) ) {
+				editables = [ selectedBlock ];
+			} else {
+				editables = Array.from(
+					selectedBlock.querySelectorAll( '[contenteditable]' )
+				);
+			}
+
+			const editableIndex = editables.indexOf( document.activeElement );
+			const selection = window.getSelection();
+
+			if ( editableIndex === -1 || ! selection.rangeCount ) {
+				return { blockIndex };
+			}
+
+			const range = selection.getRangeAt( 0 );
+			const cloneStart = range.cloneRange();
+			const cloneEnd = range.cloneRange();
+
+			cloneStart.setStart( document.activeElement, 0 );
+			cloneEnd.setStart( document.activeElement, 0 );
+
+			/**
+			 * Zero width non-breaking space, used as padding in the editable DOM
+			 * tree when it is empty otherwise.
+			 */
+			const ZWNBSP = '\ufeff';
+
+			return {
+				blockIndex,
+				editableIndex,
```

Comment:
Nit: TBH, I have no idea what these mean 😅 . Could we instead use public APIs like `wp.data.select( 'core/block-editor' ).getBlockIndex()` to compute some of these fields?

However, I think we can keep these if we can't figure out a better way to test the same thing in this PR. If it's possible, I'd like to refactor it to better match the best practices, as shown in the PR https://github.com/WordPress/gutenberg/pull/48035. I don't think that's a focus for this migration PR though!

Reply:
OK. I’ll try it.

Reply:
@kevin940726 Do you think we need `editableIndex` ?

Reply:
I don't know what it does or trying to achieve, TBH. Seems like every assertion is checking if it's equal to `0`. I'm open to delete it if you also think it doesn't help with the test.